src/hdmf/build/objectmapper.py

```diff
@@ -206,17 +216,19 @@ def convert_dtype(cls, spec, value, spec_dtype=None):  # noqa: C901
         spec_dtype_type = cls.__dtypes[spec_dtype]
         warning_msg = None
         # Numpy Array or Zarr array
-        if (isinstance(value, np.ndarray) or
-                (hasattr(value, 'astype') and hasattr(value, 'dtype'))):
+        # NOTE: Numpy < 2.0 has only fixed-length strings.
+        # Numpy 2.0 introduces variable-length strings (dtype=np.dtypes.StringDType()).
+        # HDMF does not yet do any special handling of numpy arrays with variable-length strings.
+        if isinstance(value, np.ndarray) or _is_zarr_array(value):
```

Comment:
The basic approach seems fine. A couple of thoughts:
- I think instead of the `if isinstance(value, np.ndarray) or _is_zarr_array(value)` it would be useful to split this case into `if` and `elif` cases to treat numpy and zarr separately here. I think that would be more readable. 
- I'd move the `_is_zarr_array` method ot the `utils` or `data_utils` 
- Alternatively to this approach we could also use `hasattr(value, filters)` to distinguish been Zarr and numpy arrays, because numpy arrays don't have a `filters` attribute which is used in Zarr to store all the filters, including the encoder for strings. 
-